pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3434 @@
+            return """%s
+
+        The inequality expression:
+            %s
+        contains non-constant terms (variables) that were evaluated in an
+        unexpected Boolean context at
+          File '%s', line %s%s
+
+        Evaluating Pyomo variables in a Boolean context, e.g.
+            if expression <= 5:
+        is generally invalid.  If you want to obtain the Boolean value of the
+        expression based on the current variable values, explicitly evaluate the
+        expression using the value() function:
+            if value(expression) <= 5:
+        or
+            if value(expression <= 5):
+        """ % args
+
+else:                               #pragma: no cover
+    _chainedInequality = None
+
+
+_ParamData = None
+SimpleParam = None
+TemplateExpressionError = None
+def initialize_expression_data():
+    """
+    A function used to initialize expression global data.
+
+    This function is necessary to avoid global imports.  It is executed
+    when ``pyomo.environ`` is imported.
+    """
+    global _ParamData
+    global SimpleParam
+    global TemplateExpressionError
+    from pyomo.core.base.param import _ParamData, SimpleParam
+    from pyomo.core.base.template_expr import TemplateExpressionError
+
+
+def compress_expression(expr):
+    """
+    Deprecated function that was used to compress deep Pyomo5
+    expression trees.
+    """
+    return expr
+
+
+class clone_counter_context(object):
+    """ Context manager for counting cloning events.
+
+    This context manager counts the number of times that the
+    :func:`clone_expression <pyomo.core.expr.current.clone_expression>`
+    function is executed.
+    """
+
+    _count = 0
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        pass
+
+    @property
+    def count(self):
+        """A property that returns the clone count value.
+        """
+        return clone_counter_context._count
+
+#: A clone counter context manager object that simplifies the
+#: use of this context manager.  Specifically, different
+#: instances of this context manger are not necessary.
+clone_counter = clone_counter_context()
+
+
+class mutable_sum_context(object):
+    """ Context manager for mutable sums.
+
+    This context manager is used to compute a sum while
+    treating the summation as a mutable object.
+    """
+
+    def __enter__(self):
+        self.e = _MutableSumExpression([])
+        return self.e
+
+    def __exit__(self, *args):
+        if self.e.__class__ == _MutableSumExpression:
+            self.e.__class__ = SumExpression
+
+#: A context manager object for nonlinear expressions.
+#: This is an instance of the :class:`mutable_sum_contex <pyomo.core.expr.current.mutable_sum_context>` context manager.
+#: Different instances of this context manager are not necessary, unless these context managers are nested.
+nonlinear_expression = mutable_sum_context()
```

Comment:
These should not be global, as they will cause problems with nested usage (like nested quicksum)

Reply:
Agreed